Readme.md
# twapi-ureq-rs

Twitter OAuth library used by ureq.

[Documentation](https://docs.rs/twapi-ureq)

## Features
- Application Only Authentication
- User Authentication
- Oauth1.0 Authentication
- Oauth2.0 Authentication
- JSON support(ex. dm_event, welcome_message, media_metadata, etc.)
- Multipart support(ex. post_media_upload)

## Example
```rust
use twapi_ureq::*;
use std::io::{BufReader, Cursor, Read};

fn main() {
    // Set up Environment Variables
    // CONSUMER_KEY
    // CONSUMER_SECRET
    // ACCESS_KEY
    // ACCESS_SECRET

    // OAuth2.0 Authentication
    let client = v2::Client::new_by_env().unwrap().unwrap();

    // search(Application Only Authentication)
    let res = client
        .get(
            "https://api.twitter.com/1.1/search/tweets.json",
            &vec![("q", "東京&埼玉"), ("count", "2")],
        )
        .unwrap();
    println!("{:?}", res.into_json::<serde_json::Value>());

    // OAuth1.0 Authentication
    let client = v1::Client::new_by_env().unwrap();

    // home_timeline
    let url = "https://api.twitter.com/1.1/statuses/home_timeline.json";
    let query_options = vec![("count", "2")];
    let res = client.get(url, &query_options).unwrap();
    println!("{:?}", res.into_json::<serde_json::Value>().unwrap());

    // statuses/update
    let url = "https://api.twitter.com/1.1/statuses/update.json";
    let form_options = vec![
        ("status", "!\"'#$%&\\()+,/:;<=>?@[\\]^`{|}~;-._* 全部"),
        ("in_reply_to_status_id", "1178811297455935488"),
    ];
    let res = client.post(url, &vec![], &form_options).unwrap();
    println!("{:?}", res.into_json::<serde_json::Value>());

    // direct_messages new(JSON support)
    let url = "https://api.twitter.com/1.1/direct_messages/events/new.json";
    let data = r#"{
                "event": {
                    "type": "message_create",
                    "message_create": {
                        "target": {
                            "recipient_id": "19522946"
                        },
                        "message_data": {
                            "text": "予定表〜①ﾊﾝｶｸだ!"
                        }
                    }
                }
            }"#;
    let data: serde_json::Value = serde_json::from_str(data).unwrap();
    let res = client.json(url, &vec![], data).unwrap();
    println!("{:?}", res.into_json::<serde_json::Value>());

    // media/upload(Multipart support)
    let metadata = std::fs::metadata("test.jpg").unwrap();
    let file_size = metadata.len();
    let f = std::fs::File::open("test.jpg").unwrap();
    let mut cursor = Cursor::new(vec![0; file_size as usize]);
    let mut reader = BufReader::new(f);
    reader.read(cursor.get_mut()).unwrap();

    let mut data = form::MultiPart::new();
    data.add_cursor("media", cursor);

    let url = "https://upload.twitter.com/1.1/media/upload.json";
    let res = client.multipart(url, &vec![], data).unwrap();
    println!("{:?}", res.into_json::<serde_json::Value>());
}
```

CHANGES.md
# CHANGES

## 0.2.1 - 2021-03-26

### Changed
* Update `ureq` to `2.1`


## 0.2.0 - 2021-01-07
### Changed
* Add struct `Client`
* Update `ureq` to `2.0`

## 0.1.5 - 2020-08-08
* ureq 1.0

src/v2.rs
use twapi_oauth::oauth2_authorization_header;
use ureq::{Error, Response};

pub struct Client {
    bearer_token: String,
}

impl Client {
    pub fn new(consumer_key: &str, consumer_secret: &str) -> Option<Self> {
        crate::oauth::get_bearer_token(&consumer_key, &consumer_secret)
            .map(|bearer_token| Self { bearer_token })
    }

    pub fn new_by_env() -> Result<Option<Self>, std::env::VarError> {
        Ok(Self::new(
            &std::env::var("CONSUMER_KEY")?,
            &std::env::var("CONSUMER_SECRET")?,
        ))
    }

    fn make_header(&self) -> String {
        oauth2_authorization_header(&self.bearer_token)
    }

    pub fn get(&self, url: &str, query_options: &Vec<(&str, &str)>) -> Result<Response, Error> {
        crate::raw::get(url, query_options, &self.make_header())
    }

    pub fn post(
        &self,
        url: &str,
        query_options: &Vec<(&str, &str)>,
        form_options: &Vec<(&str, &str)>,
    ) -> Result<Response, Error> {
        crate::raw::post(url, query_options, form_options, &self.make_header())
    }

    pub fn json(
        &self,
        url: &str,
        query_options: &Vec<(&str, &str)>,
        data: serde_json::Value,
    ) -> Result<Response, Error> {
        crate::raw::json(url, query_options, data, &self.make_header())
    }

    pub fn put(&self, url: &str, query_options: &Vec<(&str, &str)>) -> Result<Response, Error> {
        crate::raw::put(url, query_options, &self.make_header())
    }

    pub fn delete(&self, url: &str, query_options: &Vec<(&str, &str)>) -> Result<Response, Error> {
        crate::raw::delete(url, query_options, &self.make_header())
    }

    pub fn multipart(
        &self,
        url: &str,
        query_options: &Vec<(&str, &str)>,
        data: crate::form::MultiPart,
    ) -> Result<Response, Error> {
        crate::raw::multipart(url, query_options, data, &self.make_header())
    }
}

#[cfg(test)]
mod tests {
    use crate::v2::Client;

    #[test]
    fn test_api() {
        let client = Client::new_by_env().unwrap().unwrap();

        // search
        let res = client
            .get(
                "https://api.twitter.com/1.1/search/tweets.json",
                &vec![("q", "東京&埼玉"), ("count", "2")],
            )
            .unwrap();
        println!("{:?}", res.into_json::<serde_json::Value>());
    }
}


src/raw.rs
use ureq::{Error, Request, Response};

use twapi_oauth::encode;

fn make_query(list: &Vec<(&str, &str)>, separator: &str) -> String {
    let mut result = String::from("");
    for item in list {
        if "" != result {
            result.push_str(separator);
        }
        result.push_str(&format!("{}={}", item.0, encode(item.1)));
    }
    result
}

pub(crate) fn get(
    url: &str,
    query_options: &Vec<(&str, &str)>,
    authorization: &str,
) -> Result<Response, Error> {
    let url = if query_options.len() > 0 {
        format!("{}?{}", url, make_query(query_options, "&"))
    } else {
        url.to_owned()
    };
    ureq::get(&url).set("Authorization", authorization).call()
}

pub(crate) fn post(
    url: &str,
    query_options: &Vec<(&str, &str)>,
    form_options: &Vec<(&str, &str)>,
    authorization: &str,
) -> Result<Response, Error> {
    let request = ureq::post(url).set("Authorization", authorization).set(
        "Content-Type",
        "application/x-www-form-urlencoded;charset=UTF-8",
    );
    apply_query_options(request, query_options).send_string(&make_body(form_options))
}

pub(crate) fn json(
    url: &str,
    query_options: &Vec<(&str, &str)>,
    data: serde_json::Value,
    authorization: &str,
) -> Result<Response, Error> {
    let request = ureq::post(url)
        .set("Authorization", authorization)
        .set("Content-Type", "application/json");
    apply_query_options(request, query_options).send_json(data)
}

pub(crate) fn put(
    url: &str,
    query_options: &Vec<(&str, &str)>,
    authorization: &str,
) -> Result<Response, Error> {
    let request = ureq::put(url).set("Authorization", authorization);
    apply_query_options(request, query_options).call()
}

pub(crate) fn delete(
    url: &str,
    query_options: &Vec<(&str, &str)>,
    authorization: &str,
) -> Result<Response, Error> {
    let request = ureq::delete(url).set("Authorization", authorization);
    apply_query_options(request, query_options).call()
}

pub(crate) fn multipart(
    url: &str,
    query_options: &Vec<(&str, &str)>,
    mut data: crate::form::MultiPart,
    authorization: &str,
) -> Result<Response, Error> {
    let request = ureq::post(url).set("Authorization", authorization).set(
        "Content-Type",
        &format!("multipart/form-data; boundary={}", data.boundary),
    );
    apply_query_options(request, query_options).send_bytes(&data.to_bytes())
}

fn make_body(form_options: &Vec<(&str, &str)>) -> String {
    match serde_urlencoded::to_string(form_options) {
        Ok(body) => body
            .replace('+', "%20")
            .replace('*', "%2A")
            .replace("%7E", "~")
            .into(),
        Err(_) => String::from(""),
    }
}

fn apply_query_options(request: Request, query_options: &Vec<(&str, &str)>) -> Request {
    let mut req = request;
    for query_option in query_options {
        req = req.query(query_option.0, query_option.1);
    }
    req
}


src/form.rs
use bytebuffer::ByteBuffer;
use rand::seq::SliceRandom;
use std::io::Cursor;

pub struct MultiPart {
    pub boundary: String,
    buffer: ByteBuffer,
    before_flag: bool,
}

impl MultiPart {
    pub fn new() -> Self {
        let mut rng = &mut rand::thread_rng();
        let postfix = String::from_utf8(
            "-_ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789"
                .as_bytes()
                .choose_multiple(&mut rng, 30)
                .cloned()
                .collect(),
        )
        .unwrap();
        Self {
            boundary: format!("---------------------------{}", postfix),
            buffer: ByteBuffer::new(),
            before_flag: true,
        }
    }

    pub fn to_bytes(&mut self) -> Vec<u8> {
        if self.before_flag {
            self.write_boundary(true);
            self.before_flag = false;
        }
        self.buffer.to_bytes()
    }

    pub fn add_string(&mut self, name: &str, data: &str) {
        self.write_boundary(false);
        self.write_header(name);
        self.buffer.write_bytes(&format!("{}\r\n", data).as_bytes());
    }

    pub fn add_cursor(&mut self, name: &str, cursor: Cursor<Vec<u8>>) {
        self.write_boundary(false);
        self.write_header(name);
        self.buffer.write_bytes(cursor.get_ref());
        self.buffer.write_bytes("\r\n".as_bytes());
    }

    fn write_header(&mut self, name: &str) {
        self.buffer.write_bytes(
            &format!("Content-Disposition: form-data; name=\"{}\"\r\n\r\n", name).as_bytes(),
        );
    }

    fn write_boundary(&mut self, last_flag: bool) {
        self.buffer.write_bytes("--".as_bytes());
        self.buffer.write_bytes(&self.boundary.as_bytes());
        if last_flag {
            self.buffer.write_bytes("--".as_bytes());
        } else {
            self.buffer.write_bytes("\r\n".as_bytes());
        }
    }
}


src/oauth.rs
use std::collections::HashMap;
use twapi_oauth::calc_oauth_header;
use ureq::{Error, Response};

pub fn get_bearer_token_response(
    consumer_key: &str,
    consumer_secret: &str,
) -> Result<Response, Error> {
    let key = base64::encode(&format!("{}:{}", consumer_key, consumer_secret));
    ureq::post("https://api.twitter.com/oauth2/token")
        .set(
            "Content-Type",
            "application/x-www-form-urlencoded;charset=UTF-8",
        )
        .set("Authorization", &format!("Basic {}", key))
        .send_string("grant_type=client_credentials")
}

pub fn get_bearer_token(consumer_key: &str, consumer_secret: &str) -> Option<String> {
    match get_bearer_token_response(consumer_key, consumer_secret) {
        Ok(response) => match response.into_json::<serde_json::Value>() {
            Ok(json) => match json["access_token"].as_str() {
                Some(access_token) => Some(access_token.to_string()),
                None => None,
            },
            Err(_) => None,
        },
        Err(_) => None,
    }
}

pub fn request_token_response(
    consumer_key: &str,
    consumer_secret: &str,
    oauth_callback: &str,
    x_auth_access_type: Option<&str>,
) -> Result<Response, Error> {
    let uri = "https://api.twitter.com/oauth/request_token";
    let mut header_options = vec![("oauth_callback", oauth_callback)];
    if let Some(x_auth_access_type) = x_auth_access_type {
        header_options.push(("x_auth_access_type", x_auth_access_type));
    }
    let signed = calc_oauth_header(
        &format!("{}&", consumer_secret),
        consumer_key,
        &header_options,
        "POST",
        uri,
        &vec![],
    );
    ureq::post(uri)
        .set("Authorization", &format!("OAuth {}", signed))
        .call()
}

pub fn request_token(
    consumer_key: &str,
    consumer_secret: &str,
    oauth_callback: &str,
    x_auth_access_type: Option<&str>,
) -> Result<HashMap<String, String>, Error> {
    let response = request_token_response(
        consumer_key,
        consumer_secret,
        oauth_callback,
        x_auth_access_type,
    )?;
    Ok(parse_oauth_body(response))
}

pub fn access_token_response(
    consumer_key: &str,
    consumer_secret: &str,
    oauth_token: &str,
    oauth_token_secret: &str,
    oauth_verifier: &str,
) -> Result<Response, Error> {
    let uri = "https://api.twitter.com/oauth/access_token";
    let signed = calc_oauth_header(
        &format!("{}&{}", consumer_secret, oauth_token_secret),
        consumer_key,
        &vec![
            ("oauth_token", oauth_token),
            ("oauth_verifier", oauth_verifier),
        ],
        "POST",
        uri,
        &vec![],
    );
    ureq::post(uri)
        .set("Authorization", &format!("OAuth {}", signed))
        .call()
}

pub fn access_token(
    consumer_key: &str,
    consumer_secret: &str,
    oauth_token: &str,
    oauth_token_secret: &str,
    oauth_verifier: &str,
) -> Result<HashMap<String, String>, Error> {
    let response = access_token_response(
        consumer_key,
        consumer_secret,
        oauth_token,
        oauth_token_secret,
        oauth_verifier,
    )?;
    Ok(parse_oauth_body(response))
}

fn parse_oauth_body(response: Response) -> HashMap<String, String> {
    let mut result = HashMap::new();
    match response.into_string() {
        Ok(body) => {
            for item in body.split("&") {
                let mut pair = item.split("=");
                result.insert(
                    pair.next().unwrap().to_string(),
                    pair.next().unwrap().to_string(),
                );
            }
        }
        Err(_) => {}
    }
    result
}


src/v1.rs
use twapi_oauth::oauth1_authorization_header;
use ureq::{Error, Response};

pub struct Client {
    consumer_key: String,
    consumer_secret: String,
    access_key: String,
    access_secret: String,
}

impl Client {
    pub fn new(
        consumer_key: &str,
        consumer_secret: &str,
        access_key: &str,
        access_secret: &str,
    ) -> Self {
        Self {
            consumer_key: consumer_key.to_owned(),
            consumer_secret: consumer_secret.to_owned(),
            access_key: access_key.to_owned(),
            access_secret: access_secret.to_owned(),
        }
    }

    pub fn new_by_env() -> Result<Self, std::env::VarError> {
        Ok(Self {
            consumer_key: std::env::var("CONSUMER_KEY")?,
            consumer_secret: std::env::var("CONSUMER_SECRET")?,
            access_key: std::env::var("ACCESS_KEY")?,
            access_secret: std::env::var("ACCESS_SECRET")?,
        })
    }

    fn calc_oauth(&self, method: &str, url: &str, query_options: &Vec<(&str, &str)>) -> String {
        oauth1_authorization_header(
            &self.consumer_key,
            &self.consumer_secret,
            &self.access_key,
            &self.access_secret,
            method,
            url,
            &query_options,
        )
    }

    pub fn get(&self, url: &str, query_options: &Vec<(&str, &str)>) -> Result<Response, Error> {
        crate::raw::get(
            url,
            query_options,
            &self.calc_oauth("GET", url, &query_options),
        )
    }

    pub fn post(
        &self,
        url: &str,
        query_options: &Vec<(&str, &str)>,
        form_options: &Vec<(&str, &str)>,
    ) -> Result<Response, Error> {
        let mut merged_options = query_options.clone();
        for option in form_options {
            merged_options.push(*option);
        }
        crate::raw::post(
            url,
            query_options,
            form_options,
            &self.calc_oauth("POST", url, &merged_options),
        )
    }

    pub fn json(
        &self,
        url: &str,
        query_options: &Vec<(&str, &str)>,
        data: serde_json::Value,
    ) -> Result<Response, Error> {
        crate::raw::json(
            url,
            query_options,
            data,
            &self.calc_oauth("POST", url, &query_options),
        )
    }

    pub fn put(&self, url: &str, query_options: &Vec<(&str, &str)>) -> Result<Response, Error> {
        crate::raw::put(
            url,
            query_options,
            &self.calc_oauth("PUT", url, &query_options),
        )
    }

    pub fn delete(&self, url: &str, query_options: &Vec<(&str, &str)>) -> Result<Response, Error> {
        crate::raw::delete(
            url,
            query_options,
            &self.calc_oauth("DELETE", url, &query_options),
        )
    }

    pub fn multipart(
        &self,
        url: &str,
        query_options: &Vec<(&str, &str)>,
        data: crate::form::MultiPart,
    ) -> Result<Response, Error> {
        crate::raw::multipart(
            url,
            query_options,
            data,
            &self.calc_oauth("POST", url, &query_options),
        )
    }
}

#[cfg(test)]
mod tests {
    use crate::{form, v1::Client};
    use std::io::{BufReader, Cursor, Read};

    #[test]
    fn test_api() {
        let client = Client::new_by_env().unwrap();

        // search
        let url = "https://api.twitter.com/1.1/search/tweets.json";
        let query_options = vec![("q", "*abc"), ("count", "2")];
        let res = client.get(url, &query_options).unwrap();
        println!("{:?}", res.into_json::<serde_json::Value>().unwrap());

        // home_timeline
        let url = "https://api.twitter.com/1.1/statuses/home_timeline.json";
        let query_options = vec![("count", "2")];
        let res = client.get(url, &query_options).unwrap();
        println!("{:?}", res.into_json::<serde_json::Value>());

        // statuses/update
        let url = "https://api.twitter.com/1.1/statuses/update.json";
        let form_options = vec![
            ("status", "!\"'#$%&\\()+,/:;<=>?@[\\]^`{|}~;-._* 全部"),
            ("in_reply_to_status_id", "1178811297455935488"),
        ];
        let res = client.post(url, &vec![], &form_options).unwrap();
        println!("{:?}", res.into_json::<serde_json::Value>());

        // direct_messages new
        let url = "https://api.twitter.com/1.1/direct_messages/events/new.json";
        let data = r#"{
                    "event": {
                        "type": "message_create",
                        "message_create": {
                            "target": {
                                "recipient_id": "19522946"
                            },
                            "message_data": {
                                "text": "予定表〜①ﾊﾝｶｸだ!"
                            }
                        }
                    }
                }"#;
        let data: serde_json::Value = serde_json::from_str(data).unwrap();
        let res = client.json(url, &vec![], data).unwrap();
        println!("{:?}", res.into_json::<serde_json::Value>());

        // media/upload
        let metadata = std::fs::metadata("test.jpg").unwrap();
        let file_size = metadata.len();
        let f = std::fs::File::open("test.jpg").unwrap();
        let mut cursor = Cursor::new(vec![0; file_size as usize]);
        let mut reader = BufReader::new(f);
        reader.read(cursor.get_mut()).unwrap();

        let mut data = form::MultiPart::new();
        data.add_cursor("media", cursor);

        let url = "https://upload.twitter.com/1.1/media/upload.json";
        let res = client.multipart(url, &vec![], data).unwrap();
        println!("{:?}", res.into_json::<serde_json::Value>());
    }
}
